lib/rules/require-var-decl-first.js

```diff
@@ -0,0 +1,179 @@
+ * var x = 1;
+ * var y = 2;
+ * ```
+ * ```js
+ * var x = 1;
+ * // comments
+ * var y = 2;
+ * ```
+ * ```js
+ * var x = 1;
+ * // comments
+ * // comments 2
+ * var y = 2;
+ * ```
+ * ```js
+ * const a = 1;
+ * const b = 2;
+ * ```
+ * ```js
+ * var x = 1;
+ * function y() {var z;};
+ * ```
+ * ```js
+ * var x = 1;
+ * var y = function () {var z;};
+ * ```
+ * ```js
+ * var w = 1;
+ * function x() {
+ *  var y;
+ * // comments
+ * // comments 2
+ *  var z;
+ * };
+ * ```
+ * ```js
+ * var x = 1;
+ * var y;
+ * for (y = 0; y < 10; y++) {};
+ * ```
+ *
+ * ##### Invalid
+ *
+ * ```js
+ * var x;
+ * x = 1;
+ * var y = 2;
+ * ```
+ * ```js
+ * var w = 1;
+ * function x() {var y;};
+ * var z = 2;
+ * ```
+ * ```js
+ * var w = 1;
+ * function x() {
+ *  var y;
+ *  y = 2;
+ *  var z;
+ * };
+ * ```
+ * ```js
+ * var a;
+ * for(var count=0;count < 10;count++){}
+ * ```
+  * ```js
+ * var x;
+ * for(var count=0;count < 10;count++){
+ *  var y;
+ * }
+ * ```
+ *
+ */
+
+var assert = require('assert');
+
+module.exports = function() {};
+
+module.exports.prototype = {
+    configure: function(options) {
+        assert(
+            options === true,
+            this.getOptionName() + ' option requires a true value'
+        );
+    },
+
+    getOptionName: function() {
+        return 'requireVarDeclFirst';
+    },
+
+    check: function(file, errors) {
+        var parentNodes = {};
+        var commentTokens = [];
+
+        file.iterateTokensByType(['Line', 'Block'], function(token) {
+            commentTokens.push(token);
+        });
+
+        file.iterateNodesByType(['VariableDeclaration'], function(node) {
+            var count;
+            var comment;
+            var commentLength;
+            var parentNode;
+            var parentNodeStack;
+            var lastNode;
+            var whiteSpaceLength;
+
+            parentNode = node.parentNode;
+            while (parentNode.type !== 'Program' &&
+                parentNode.type !== 'FunctionDeclaration' &&
+                parentNode.type !== 'FunctionExpression') {
+                parentNode = parentNode.parentNode;
+            }
+
+            if (!parentNodes.hasOwnProperty(parentNode.range[0])) {
+                parentNodes[parentNode.range[0]] = { hasNonVarDecl: false, declarations: [] };
+            }
+            parentNodeStack = parentNodes[parentNode.range[0]];
+
+            if (parentNodeStack.declarations.length === 0) {
+                parentNodeStack.declarations.push(node);
+            } else {
+                lastNode = parentNodeStack.declarations[parentNodeStack.declarations.length - 1];
+                parentNodeStack.declarations.push(node);
+                if (!parentNodeStack.hasNonVarDecl) {
+                    if (node.range[0] <= lastNode.range[1]) {
+                        return;
+                    }
+
+                    whiteSpaceLength = file.getFirstNodeToken(node).whitespaceBefore.length;
+                    commentLength = 0;
```

Comment:
For clarity and reducing the amount of code in this callback, please extract the computation of this to a function.